com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Builder.java

```diff
@@ -46,13 +54,94 @@ public BuilderParams getParams() {
      * @param input input resource
      * @return new task with single input/output
      */
-    protected Task<T> defaultTask(IResource input) {
-        Task<T> task = Task.<T>newBuilder(this)
+    protected Task<T> defaultTask(IResource input) throws CompileExceptionError, IOException {
+        Task.TaskBuilder<T> taskBuilder = Task.<T>newBuilder(this)
                 .setName(params.name())
                 .addInput(input)
-                .addOutput(input.changeExt(params.outExt()))
-                .build();
-        return task;
+                .addOutput(input.changeExt(params.outExt()));
+        return taskBuilder.build();
+    }
+
+    /**
+     * Create a task whose outputs will be added to the current task's inputs.
+     * @param input input resource
+     * @param builderClass class to build resource with
+     * @param builder current task builder
+     * @return new subtask with single input/output
+     */
+    protected Task<?> createSubTask(IResource input, Class<? extends Builder<?>> builderClass, Task.TaskBuilder<?> builder) throws CompileExceptionError {
+        Task<?> subTask = project.createTask(input, builderClass);
+        builder.addInputsFromOutputs(subTask);
+        return subTask;
+    }
+
+    /**
+     * Create a task whose outputs will be added to the current task's inputs.
+     * @param input input resource
+     * @param builder current task builder
+     * @return new subtask with single input/output
+     */
+    protected Task<?> createSubTask(IResource input, Task.TaskBuilder<?> builder) throws CompileExceptionError {
+        Task<?> subTask = project.createTask(input);
+        if (subTask == null) {
+            throw new CompileExceptionError(input,
+                    0,
+                    String.format("Failed to create build task for '%s'", input.getPath()));
+        }
+        builder.addInputsFromOutputs(subTask);
+        return subTask;
+    }
+
+    /**
+     * Create a task whose outputs will be added to the current task's inputs.
+     * @param inputPath input path to resource
+     * @param field where specified path to the resource
+     * @param builder current task builder
+     * @return new subtask with single input/output
+     */
+    protected Task<?> createSubTask(String inputPath, String field, Task.TaskBuilder<?> builder) throws CompileExceptionError {
+        IResource res = BuilderUtil.checkResource(project, builder.firstInput(), field, inputPath);
+        Task<?> subTask = project.createTask(res);
+        builder.addInputsFromOutputs(subTask);
+        return subTask;
+    }
+
+    protected void createSubTasks(MessageOrBuilder builder, Task.TaskBuilder<T> taskBuilder) throws CompileExceptionError {
```

Comment:
Here we scan fields of the builder (or message) and create tasks all the `isResources` except optional fields (see the next comment)